### ISSUE TITLE
Failed to compile 05-geometry-operations.Rmd

### ISSUE THREAD

**Issue by phungdo** [Robinlovelace/geocompr#508]
When I compile the book. I seems to I can not compile since chapter 5. Could you take a look of the warning I copy from the terminal.

> Quitting from lines 851-863 (05-geometry-operations.Rmd) 
> **Error in CPL_transform(x, crs, aoi, pipeline, reverse)** : 
>   OGRCreateCoordinateTransformation() returned NULL: PROJ available?
> Calls: local ... st_transform -> st_transform.sfc -> structure -> CPL_transform
> In addition: There were 11 warnings (use warnings() to see them)
> 
> Execution halted
> Error in Rscript_render(f, render_args, render_meta, add1, add2) : 
>   Failed to compile 05-geometry-operations.Rmd

**Comment by Robinlovelace** [comment]
It's most likely an issue with the version of `sf` and / or PROJ you have installed. Can you try running it on a Docker container, e.g. as follows?:

```
docker run -d -p 8783:8787 -v ~/atfutures/:/home/rstudio/data -e PASSWORD=<strongpassword >
```

**Comment by phungdo** [comment]
> It's most likely an issue with the version of `sf` and / or PROJ you have installed. Can you try running it on a Docker container, e.g. as follows?:
> 
> ```
> docker run -d -p 8783:8787 -v ~/atfutures/:/home/rstudio/data -e PASSWORD=<strongpassword >
> ```

I got the **library(sf)** and also PROJ 0.1.6 installed with 

> Linking to GEOS 3.6.1, GDAL 2.2.3, PROJ 4.9.3

I have not tried via Docker.

**Comment by Robinlovelace** [comment]
> If possible can you add a button on your webpage to automatically generate PDF file.

We deliberately do not distribute PDF versions

1. Because the book's contents are changing and PDF versions risk spreading out-of-date information, and
2. Although the chapters are free to read online, we want people to [buy physical copies](https://www.amazon.com/Geocomputation-Chapman-Hall-Robin-Lovelace/dp/1138304514/) of the book to make open source book projects like this viable for publishers such as CRC Press.

**Comment by mattwilliamson13** [comment]
I am also having this issue (same error, same part of Ch. 5), but on a Mac (Catalina OS). ```sf```, ```gdal```, and ```proj``` were all updated according to the instructions on the ```sf```.  Just alerting you to this in the event that this issue should remain open.

**Comment by Robinlovelace** [comment]
Thanks for reporting this @mattwilliamson13. Which versions of GDAL, GEOS and PROJ do you have? You could report that by running the following command:


```r
reprex::reprex(library(sf))
```

That generates the following output:

``` r
library(sf)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
```

<sup>Created on 2020-04-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by Robinlovelace** [comment]
Also, could you try running the section that is crashing in the browser with this link: https://mybinder.org/v2/gh/robinlovelace/geocompr/master?urlpath=rstudio 

If it runs on one system and not another that could help identify the underlying cause. Another option is to try re-installing spData, spDataLarge and geocompkg. Interested to find out the underlying cause, let us know how you get on.

**Comment by mattwilliamson13** [comment]
> Thanks for reporting this @mattwilliamson13. Which versions of GDAL, GEOS and PROJ do you have? You could report that by running the following command:
> 
> ```r
> reprex::reprex(library(sf))
> ```
> 
> That generates the following output:
> 
> ```r
> library(sf)
> #> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
> ```
> 
> Created on 2020-04-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)
Hmmm. This is the output (suggesting we are running different versions). I followed the home-brew installation instructions on the ```sf``` webpage, so not sure why my versions are so far behind yours?
library(sf)
#> Linking to GEOS 3.7.2, GDAL 2.4.2, PROJ 5.2.0

**Comment by Robinlovelace** [comment]
> Hmmm. This is the output (suggesting we are running different versions). I followed the home-brew installation instructions on the sf webpage, so not sure why my versions are so far behind yours?

```r
library(sf)
#> Linking to GEOS 3.7.2, GDAL 2.4.2, PROJ 5.2.0
```

I think that is the underlying issue @mattwilliamson13. If you try to install the latest versions that may help, but I'm not a Mac user so cannot help, other than to point you to resources such as this: https://github.com/r-spatial/sf/blob/master/.github/workflows/R-CMD-check.yaml#L45

**Comment by Robinlovelace** [comment]
Seems possible to install more recent versions of PROJ on Mac: https://github.com/r-spatial/sf/runs/594736641?check_suite_focus=true#step:11:80

**Comment by mattwilliamson13** [comment]
Ok - the ```binder``` instance works, I'm not quite sure what to do with the link that you sent (they seem to be package checks, but not sure how to make those work locally)

**Comment by mattwilliamson13** [comment]
Follow up. Installing the development version of ```sf``` and then running the ```reprex::reprex(library(sf))``` yields the same response as the one you indicated. I then start a new Studio session, switch to the ```geocompr``` project and run
```bookdown::render_book("index.Rmd")```
which returns the following:
```pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
Quitting from lines 851-863 (05-geometry-operations.Rmd) 
Error in CPL_transform(x, crs, aoi, pipeline, reverse) : 
  OGRCreateCoordinateTransformation() returned NULL: PROJ available?
Calls: local ... st_transform -> st_transform.sfc -> structure -> CPL_transform
In addition: Warning message:
In CPL_transform(x, crs, aoi, pipeline, reverse) :
  GDAL Error 1: No PROJ.4 translation for source SRS, coordinate transformation initialization has failed.

Execution halted
Error in Rscript_render(f, render_args, render_meta, add1, add2) : 
  Failed to compile 05-geometry-operations.Rmd```

This is a slightly different message, but seems to indicate a similar issues?

**Comment by Nowosad** [comment]
@mattwilliamson13, what is your GDAL and PROJ versions?

**Comment by Nowosad** [comment]
@mattwilliamson13 Please try to update **spDataLarge** and let us know if it works now.
```r
install.packages("spDataLarge", repos = "https://nowosad.github.io/drat/", type = "source")
```

**Comment by mattwilliamson13** [comment]
GDAL and PROJ are: 
``` r
library(sf)
#> Linking to GEOS 3.8.1, GDAL 2.4.4, PROJ 7.0.0
```

<sup>Created on 2020-04-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mattwilliamson13** [comment]
Updated spDataLarge (following the code above). Returned:
```
pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
pj_obj_create: Cannot find proj.db
Quitting from lines 851-863 (05-geometry-operations.Rmd) 
Error in CPL_transform(x, crs, aoi, pipeline, reverse) : 
  OGRCreateCoordinateTransformation() returned NULL: PROJ available?
Calls: local ... st_transform -> st_transform.sfc -> structure -> CPL_transform
In addition: Warning message:
In CPL_transform(x, crs, aoi, pipeline, reverse) :
  GDAL Error 1: No PROJ.4 translation for source SRS, coordinate transformation initialization has failed.

Execution halted
Error in Rscript_render(f, render_args, render_meta, add1, add2) : 
  Failed to compile 05-geometry-operations.Rmd
```

**Comment by mattwilliamson13** [comment]
I'm not sure if this is something happening with ```tmap```? It seems to fail at the rendering of ```rast_poly_point```

**Comment by Nowosad** [comment]
@mattwilliamson13 I have tested this code on PROJ5 and PROJ7 - and it works. I think there is something wrong with your PROJ installation - I do not have this message - `pj_obj_create: Cannot find proj.db`.

**Comment by mattwilliamson13** [comment]
I reinstalled PROJ using:
```brew install proj``` 
and then gdal using:
```
brew install pkg-config
brew install gdal
```
which returns:
```
gdal dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime
```

Is this indicating a problem with the compiler? If so, it may be related to the fact that I also use Stan, which had some issues with the latest upgrade to Catalina (https://github.com/stan-dev/rstan/wiki/Catalina-problems; and https://github.com/stan-dev/rstan/wiki/Installing-RStan-from-source-on-a-Mac). At this point, I'm a bit at a loss...

**Comment by Robinlovelace** [comment]
Update: I suspect there is an issue with the GDAL/PROJ combination and sf... Trying on Docker.

**Comment by Robinlovelace** [comment]
Update: I can reproduce similar errors in C8 on a system with UbuntuGIS stable PROJ/GDAL:

```
label: urban-facet (with options)
List of 4
$ message : logi FALSE
$ fig.cap : chr "Faceted map showing the top 30 largest urban agglomerations from 1970 to 2030 based on population projections b"| __truncated__
$ fig.scap: chr "Faceted map showing urban agglomerations."
$ fig.asp : num 0.5
[91mQuitting from lines 420-428 (08-mapping.Rmd)
[0m
[91mError in CPL_transform(x, crs, aoi, pipeline, reverse) :
OGRCreateCoordinateTransformation() returned NULL: PROJ available?
Calls: local ... st_transform -> st_transform.sfc -> structure -> CPL_transform
```

To get that system:

```
docker run -it -e DISABLE_AUTH=TRUE geocompr/geocompr:ubuntugis_stable bash
R
library(sf)
#> Linking to GEOS 3.7.1, GDAL 2.4.2, PROJ 5.2.0
library(spData)
```

**Comment by Nowosad** [comment]
@Robinlovelace have you tried updating **spDataLarge**?

```r
install.packages("spDataLarge", repos = "https://nowosad.github.io/drat/", type = "source")
```

**Comment by Robinlovelace** [comment]
Not yet...

**Comment by Robinlovelace** [comment]
Trying it now:

```
docker run -it -e DISABLE_AUTH=TRUE geocompr/geocompr:ubuntugis_stable bash
R
remotes::install_github("Nowosad/spData")
q()

cd /home/rstudio && \
  wget https://github.com/Robinlovelace/geocompr/archive/master.zip && \
  unzip master.zip && \
  mv geocompr-master /home/rstudio/geocompr && \
  cd geocompr && \
  make html
  
```

**Comment by Robinlovelace** [comment]
Interesting that it only affects the UbuntuGIS stable version...

![image](https://user-images.githubusercontent.com/1825120/79779373-3a6c4900-8332-11ea-87f3-8b4875f9f459.png)

**Comment by Robinlovelace** [comment]
> and spData?

Good news @Nowosad that seems to have fixed it. Meaning that when `spData` is update on CRAN we can close this issue?

**Comment by Nowosad** [comment]
@Robinlovelace the issue is probably due to GDAL2 and PROJ4/5. `urban_agglomeration` had its CRS created on PROJ7. However, even PROJ5 has a problem with reading it. Two days ago I changed CRS of most objects in spData and spDataLarge to PROJ5. Now, it should work on most of the machines...

**Comment by Robinlovelace** [comment]
Makes sense. Thanks for navigating the fiddly and potentially frustrating world of PROJ versions and looking forward to the latest version on CRAN. These cryptic error messages are hard to debug.

**Comment by Robinlovelace** [comment]
My guess: there was a regression in PROJ 5 that has since been fixed. But this is over my head.

**Comment by Nowosad** [comment]
I can show you the difference. 
**PROJ5**:
```
> st_crs(world)
Coordinate Reference System:
  User input: EPSG:4326 
  wkt:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.257223563,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0,
        AUTHORITY["EPSG","8901"]],
    UNIT["degree",0.0174532925199433,
        AUTHORITY["EPSG","9122"]],
    AUTHORITY["EPSG","4326"]]
```

**PROJ7**:

```
> st_crs(world)
Coordinate Reference System:
  User input: WGS 84 
  wkt:
GEOGCRS["WGS 84",
    DATUM["World Geodetic System 1984",
        ELLIPSOID["WGS 84",6378137,298.257223563,
            LENGTHUNIT["metre",1]]],
    PRIMEM["Greenwich",0,
        ANGLEUNIT["degree",0.0174532925199433]],
    CS[ellipsoidal,2],
        AXIS["geodetic latitude (Lat)",north,
            ORDER[1],
            ANGLEUNIT["degree",0.0174532925199433]],
        AXIS["geodetic longitude (Lon)",east,
            ORDER[2],
            ANGLEUNIT["degree",0.0174532925199433]],
    USAGE[
        SCOPE["unknown"],
        AREA["World"],
        BBOX[-90,-180,90,180]],
    ID["EPSG",4326]]
```

**Comment by mattwilliamson13** [comment]
I'm not entirely sure I follow this thread, but I am still unable to successfully render the book locally despite updating all packages (```spData```, ```spDataLarge```). I get:
```
Quitting from lines 851-863 (05-geometry-operations.Rmd) 
Error in CPL_transform(x, crs, aoi, pipeline, reverse) : 
  OGRCreateCoordinateTransformation() returned NULL: PROJ available?
Calls: local ... st_transform -> st_transform.sfc -> structure -> CPL_transform
In addition: Warning message:
In CPL_transform(x, crs, aoi, pipeline, reverse) :
  GDAL Error 1: No PROJ.4 translation for source SRS, coordinate transformation initialization has failed.
```
when using 
``` r
library(sf)
#> Linking to GEOS 3.7.2, GDAL 2.4.2, PROJ 5.2.0
```

<sup>Created on 2020-04-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

So, it appears that I am unable to render successfully under Proj5 or 7.

**Comment by Nowosad** [comment]
What's your sf version? Also, GDAL2 and PROJ5 is not a good combination. Try to move to GDAL3.

**Comment by Robinlovelace** [comment]
> Try to move to GDAL3.

Fully agree. You can do this with Docker, e.g. as outlined here: https://github.com/geocompr/docker#docker

**Comment by Robinlovelace** [comment]
Any update with this @phungdo? I think we can close the issue on the basis that this issue only affects certain operating systems and there is nothing that I can see immediately that we can change in the book to fix it. I think updating `sf` when the next version ships with GDAL 3.x will fix it as per https://github.com/r-spatial/sf/issues/1386#issuecomment-622792042 so closing for now. Happy to re-visit this if there are changes we can make in the book to resolve the issue but I think it's a computer set-up and upstream issue.

**Comment by phungdo** [comment]
> Any update with this @phungdo? I think we can close the issue on the basis that this issue only affects certain operating systems and there is nothing that I can see immediately that we can change in the book to fix it. I think updating `sf` when the next version ships with GDAL 3.x will fix it as per [r-spatial/sf#1386 (comment)](https://github.com/r-spatial/sf/issues/1386#issuecomment-622792042) so closing for now. Happy to re-visit this if there are changes we can make in the book to resolve the issue but I think it's a computer set-up and upstream issue.

Hi @Robinlovelace,
Thank you for mentioning me. The way of studying the online book is a better option for me. Therefore, I don't try to compile the book for offline using. If I have any problems I will make a ticket.
